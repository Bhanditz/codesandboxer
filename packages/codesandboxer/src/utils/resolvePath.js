// @flow
const rp2 = (base, rel) => {
  // this logic is designed to prevent us leaving our current scope
  switch (rel[0]) {
    case '..': {
      if (base.length < 1) {
        throw new Error(
          `invalid relative path from ${base.join('/')} to ${rel.join('/')}`,
        );
      }
      return rp2(base.slice(0, -1), rel.slice(1));
    }
    default: {
      return `${base.concat(rel).join('/')}`;
    }
  }
};

const resolvePath = (basePath: string, relativePath: string): string => {
  let base = basePath.split('/').filter(a => a);
  let rel = relativePath.split('/').filter(a => a);

  let val = '';
  switch (rel[0]) {
    case '.':
      // We are assuming that the base is always a file path
      val = base
        .slice(0, -1)
        .concat(rel.slice(1))
        .join('/');
      break;
    case '..':
      // this logic is designed to prevent us leaving our current scope
      if (base.length < 1) {
        throw new Error(
          `invalid relative path from ${basePath} to ${relativePath}`,
        );
      }
      val = rp2(base.slice(0, -1), rel);
      break;
    default:
      // TODO: Decide if this is correct or terribly wrong
      val = base.concat(rel).join('/');
  }

  return val;
};

export default resolvePath;
