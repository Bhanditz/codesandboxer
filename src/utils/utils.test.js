import cases from 'jest-in-case';
import getAllImports from './getAllImports';
import resolvePath from './resolvePath';

const codeImportTests = [
  { name: 'simple import', code: `import a from 'b'` },
  { name: 'spread import', code: `import { a } from 'b'` },
  { name: 'two imports', code: `import a from 'b' import c from 'd'` },
  {
    name: 'multiline imports',
    code: `import a from 'b'
import c from 'd'`,
  },
  { name: 'two spread imports', code: `import { a, b } from 'c'` },
  {
    name: 'two spread imports multiline',
    code: `import {
  a,
  b
} from 'c'`,
  },
  { name: 'no spaces', code: `import {a} from 'b'` },
  {
    name: 'dev and peer deps',
    code: `import {a} from 't' import s from 'z' import y from 'x'`,
  },
  {
    name: 'relativeImport',
    code: `import {a} from './c'`,
  },
  {
    name: 'using regex pattern',
    code: `import a from './c/somewhere' import b from './c/anywhere'`,
    old: './c/*',
    new: 'c/',
  },
];

cases(
  'getAllImports()',
  ({ code }) => {
    let mpts = getAllImports(code);
    expect(mpts).toMatchSnapshot();
  },
  codeImportTests,
);

cases(
  'resolvePath',
  ({ basePath, relativePath, returnedPath }) => {
    let res = resolvePath(basePath, relativePath);
    expect(res).toBe(returnedPath);
  },
  [
    { basePath: 'a/b/c', relativePath: '../z', returnedPath: 'a/z' },
    {
      basePath: 'a/b/c',
      relativePath: '../../z/x',
      returnedPath: 'z/x',
    },
    { basePath: 'a/b/c', relativePath: './z', returnedPath: 'a/b/z' },
    {
      basePath: 'a/b/c',
      relativePath: '../../../z/x',
      returnedPath: 'a/b/c',
    },
    {
      basePath: 'a/b/c',
      relativePath: 'zxy',
      returnedPath: 'a/b/c/zxy',
    },
    {
      basePath: 'a/b/c/',
      relativePath: './zxy',
      returnedPath: 'a/b/zxy',
    },
  ],
);
