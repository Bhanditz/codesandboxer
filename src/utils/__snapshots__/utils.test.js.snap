// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllImports() dev and peer deps 1`] = `
Array [
  Array [
    "import {a} from 't'",
    "t",
  ],
  Array [
    "import s from 'z'",
    "z",
  ],
  Array [
    "import y from 'x'",
    "x",
  ],
]
`;

exports[`getAllImports() multiline imports 1`] = `
Array [
  Array [
    "import a from 'b'",
    "b",
  ],
  Array [
    "import c from 'd'",
    "d",
  ],
]
`;

exports[`getAllImports() no spaces 1`] = `
Array [
  Array [
    "import {a} from 'b'",
    "b",
  ],
]
`;

exports[`getAllImports() relativeImport 1`] = `
Array [
  Array [
    "import {a} from './c'",
    "./c",
  ],
]
`;

exports[`getAllImports() simple import 1`] = `
Array [
  Array [
    "import a from 'b'",
    "b",
  ],
]
`;

exports[`getAllImports() spread import 1`] = `
Array [
  Array [
    "import { a } from 'b'",
    "b",
  ],
]
`;

exports[`getAllImports() two imports 1`] = `
Array [
  Array [
    "import a from 'b'",
    "b",
  ],
  Array [
    "import c from 'd'",
    "d",
  ],
]
`;

exports[`getAllImports() two spread imports 1`] = `
Array [
  Array [
    "import { a, b } from 'c'",
    "c",
  ],
]
`;

exports[`getAllImports() two spread imports multiline 1`] = `
Array [
  Array [
    "import {
  a,
  b
} from 'c'",
    "c",
  ],
]
`;

exports[`getAllImports() using regex pattern 1`] = `
Array [
  Array [
    "import a from './c/somewhere'",
    "./c/somewhere",
  ],
  Array [
    "import b from './c/anywhere'",
    "./c/anywhere",
  ],
]
`;
